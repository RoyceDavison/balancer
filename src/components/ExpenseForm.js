import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
// import "react-dates/lib/css/_datepicker.css";
import "react-dates/initialize";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.onDescriptionChange = this.onDescriptionChange.bind(this);
    this.onNoteChange = this.onNoteChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);

    this.state = {
      des: props.expense ? props.expense.des : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onDescriptionChange(e) {
    const description = e.target.value;
    this.setState(() => {
      return {
        des: description,
      };
    });
  }

  onNoteChange(e) {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  onAmountChange(e) {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount,
      }));
    }
  }

  onDateChange(createdAt) {
    //prevent the user to enter "DELETE" to delete the data
    if (createdAt) {
      this.setState(() => ({
        createdAt,
      }));
    }
  }

  onFocusChange({ focused }) {
    this.setState(() => ({
      calendarFocused: focused,
    }));
  }

  onSubmit(e) {
    e.preventDefault();
    if (!this.state.des || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount.",
      }));
    } else {
      this.setState(() => ({
        error: "",
      }));

      //pass down to AddExpensePage
      this.props.onSubmit({
        des: this.state.des,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(), //from moment() library,
        note: this.state.note,
      });
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <h3>{this.state.error}</h3>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.des}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            id="datePicker"
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={(day) => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense (optional)"
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
