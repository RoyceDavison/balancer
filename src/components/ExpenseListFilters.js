import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../redux/actions/filters";

class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.onSortChange = this.onSortChange.bind(this);

    this.state = {
      calendarFocused: null,
    };
  }

  //look at github react-dates then we could find the returned parameters
  onDatesChange({ startDate, endDate }) {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  }

  onFocusChange(calendarFocused) {
    this.setState(() => ({
      calendarFocused,
    }));
  }

  onTextChange(e) {
    this.props.dispatch(setTextFilter(e.target.value));
  }

  onSortChange(e) {
    if (e.target.value === "date") {
      this.props.dispatch(sortByDate());
    } else if (e.target.value === "amount") {
      this.props.dispatch(sortByAmount());
    }
    // console.log(e.target.value);
  }

  render() {
    return (
      //controled input
      <div>
        <input
          type="text"
          defaultValue=""
          //value={props.filters.text}
          onChange={this.onTextChange}
        />

        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          startDateId="your_unique_start_date_id"
          endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          endDateId="your_unique_end_date_id"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);

// const ExpenseListFilters = (props) => (
//   //controled input
//   <div>
//     <input
//       type="text"
//       defaultValue=""
//       //value={props.filters.text}
//       onChange={(e) => {
//         props.dispatch(setTextFilter(e.target.value));
//       }}
//     />

//     <select
//       value={props.filters.sortBy}
//       onChange={(e) => {
//         if (e.target.value === "date") {
//           props.dispatch(sortByDate());
//         } else if (e.target.value === "amount") {
//           props.dispatch(sortByAmount());
//         }
//         // console.log(e.target.value);
//       }}
//     >
//       <option value="date">Date</option>
//       <option value="amount">Amount</option>
//     </select>
//   </div>
// );
