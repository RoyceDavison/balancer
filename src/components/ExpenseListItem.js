import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../redux/actions/expenses";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ dispatch, id, des, amount, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{des}</h3>
      <span className="list-item__sub-title">
        {moment(createdAt).format("MMMM Do, YYYY")}
      </span>
    </div>
    <h3 className="list-item__data">
      {numeral(amount / 100).format("$0,0.00")}
    </h3>
  </Link>

  /*{ <button
      onClick={() => {
        dispatch(removeExpense({ id }));
      }}
    >
      Remove
    </button> }*/
);

// export default connect()(ExpenseListItem);
export default ExpenseListItem;
