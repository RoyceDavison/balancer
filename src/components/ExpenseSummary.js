import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectExpenses from "../redux/selectors/expenses";
import selectExpensesTotal from "../redux/selectors/expenses-total";

export const ExpenseSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount == 1 ? "expense" : "expenses";
  const formattedExpenseTotal = numeral(expenseTotal / 100).format("$0,0.00");
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpenseTotal}
      </h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpense = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpense.length,
    expenseTotal: selectExpensesTotal(visibleExpense),
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
