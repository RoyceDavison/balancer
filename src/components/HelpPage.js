import React from "react";

const HelpPage = () => <div>This is React help page</div>;

export default HelpPage;
