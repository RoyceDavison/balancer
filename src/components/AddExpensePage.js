import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../redux/actions/expenses";

export const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        //props.dispatch(addExpense(expense));

        //for testing
        props.onSubmit(expense);
        props.history.push("/");
      }}
    />
  </div>
);

//use for testing purpose
const mapDispatchProps = (dispatch) => {
  return {
    onSubmit: (expense) => dispatch(addExpense(expense)),
  };
};

export default connect(undefined, mapDispatchProps)(AddExpensePage);
