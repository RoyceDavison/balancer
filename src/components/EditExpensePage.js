import React, { Children } from "react";
import { connect } from "react-redux";
import ExpenseForm from "../components/ExpenseForm";
import {
  startEditExpense,
  startRemoveExpense,
} from "../redux/actions/expenses";

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onRemove = this.onRemove.bind(this);
  }

  onSubmit(expense) {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  }

  onRemove() {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        {<button onClick={this.onRemove}>Remove</button>}
      </div>
    );
  }
}

const mapDispatchProps = (dispatch, props) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
  };
};

const mapStateToProp = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id;
    }),
  };
};

export default connect(mapStateToProp, mapDispatchProps)(EditExpensePage);
