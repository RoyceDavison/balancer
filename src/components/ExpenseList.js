import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../redux/selectors/expenses";

//connect a component to the redux store, it is reactive: when the store change, the component get re-render
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>
            <h3>No Expense</h3>
          </span>
        </div>
      ) : (
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )}
    </div>
  </div>
);

//HOC:higer order component
//Version 1
// const ConnectExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses,
//   };
// })(ExpenseList);
// export default ConnectExpenseList;

//Version 2
// export default connect((state) => {
//   return {
//     expenses: state.expenses,
//   };
// })(ExpenseList);

//Version 3
//mapStateToProps is used to defined the things we want to get off from the store
// const mapStateToProps = (state) => {
//   return {
//     expenses: state.expenses,
//     filters: state.filters,
//   };
// };

//mapStateToProps is used to defined the things we want to get off from the store
//so the regular component could access these things via props
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
