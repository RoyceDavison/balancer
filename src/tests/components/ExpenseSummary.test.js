import React from "react";
import { shallow } from "enzyme";
import { ExpenseSummary } from "../../components/ExpenseSummary";

test("Should correctly render ExpenseSummary with 1 expense", () => {
  const wrapper = shallow(
    <ExpenseSummary expenseCount={1} expenseTotal={235} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("Should correctly render ExpenseSummary with multiple expenses", () => {
  const wrapper = shallow(
    <ExpenseSummary expenseCount={23} expenseTotal={320984930} />
  );
  expect(wrapper).toMatchSnapshot();
});
