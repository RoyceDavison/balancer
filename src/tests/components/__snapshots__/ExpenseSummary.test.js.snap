// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpenseSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 3 
    1
     
    expense
     totalling 
    $235.00
  </h1>
</div>
`;

exports[`Should correctly render ExpenseSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 3 
    23
     
    expenses
     totalling 
    $320,984,930.00
  </h1>
</div>
`;
